projects/ssr-tests/src/ssr-testing.spec.ts

```diff
@@ -1,54 +1,126 @@
-import * as Log from './log.utils';
-import * as ProxyServer from './proxy.utils';
-import * as Ssr from './ssr.utils';
+import { Server } from 'http';
+import { expectLogMessages } from './matchers/matchers';
+import * as LogUtils from './utils/log.utils';
+import * as ProxyUtils from './utils/proxy.utils';
+import * as SsrUtils from './utils/ssr.utils';
 
 const BACKEND_BASE_URL: string = process.env.CX_BASE_URL || '';
 
 describe('SSR E2E', () => {
-  let proxy: any;
+  let proxy: Server;
   const REQUEST_PATH = '/electronics-spa/en/USD/';
 
-  beforeEach(async () => {
-    Log.clearSsrLogFile();
-    await Ssr.startSsrServer();
+  beforeEach(() => {
+    LogUtils.clearSsrLogFile();
   });
 
   afterEach(async () => {
-    await proxy.close();
-    await Ssr.killSsrServer();
+    proxy.close();
+    await SsrUtils.killSsrServer();
   });
 
-  it('should receive success response with request', async () => {
-    proxy = await ProxyServer.startProxyServer({
-      target: BACKEND_BASE_URL,
-    });
-    const response: any = await ProxyServer.sendRequest(REQUEST_PATH);
-    expect(response.statusCode).toEqual(200);
-
-    // Rendering should not complete in the first request.
-    // App should fall back to csr.
-    Log.assertMessages([
-      `Rendering started (${REQUEST_PATH})`,
-      `Request is waiting for the SSR rendering to complete (${REQUEST_PATH})`,
-    ]);
-  });
+  describe('With SSR error handling', () => {
+    describe('Common behavior', () => {
+      beforeEach(async () => {
+        await SsrUtils.startSsrServer();
+      });
+
+      it('should receive success response with request', async () => {
+        proxy = await ProxyUtils.startBackendProxyServer({
+          target: BACKEND_BASE_URL,
+        });
+        const response: any =
+          await ProxyUtils.sendRequestToSsrServer(REQUEST_PATH);
+        expect(response.statusCode).toEqual(200);
+
+        // Rendering should not complete in the first request.
+        // App should fall back to csr.
+        expectLogMessages().toContainLogs([
+          `Rendering started (${REQUEST_PATH})`,
+          `Request is waiting for the SSR rendering to complete (${REQUEST_PATH})`,
+        ]);
+      });
+
+      it('should receive response with 404 when page does not exist', async () => {
+        proxy = await ProxyUtils.startBackendProxyServer({
+          target: BACKEND_BASE_URL,
+        });
+        const response = await ProxyUtils.sendRequestToSsrServer(
+          REQUEST_PATH + 'not-existing-page'
+        );
+        expect(response.statusCode).toEqual(404);
```

Comment:
[MINOR]
What about adding an `expect` that we want to contain logs:
```
`Request is resolved with the SSR rendering error (${REQUEST_PATH + 'not-existing-page'})`
```
?

Reply:
For some reasons, tests ends before this exact log is generated and written to `.ssr.log` file from which we're taking log messages for assertion purposes. I haven't found a reason of this, but as it's rather framework issue, I don't want to spend more time on it. IMO the test can be adjusted later.